### PR TITLE
🎉 New source: DynamoDB 

### PR DESCRIPTION
**Pull request by Tomperez98**
## What
Missing `source-dynamodb`. WIP

## How
Implementing `source-dynamodb`

## Recommended reading order
`y.python`

## 🚨 User Impact 🚨
More source optionality

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [X] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

```
Name                          Stmts   Miss  Cover
-------------------------------------------------
source_dynamodb/__init__.py       2      0   100%
source_dynamodb/reader.py        64     12    81%
source_dynamodb/source.py        36     16    56%
-------------------------------------------------
TOTAL                           102     28    73%
```

</details>

<details><summary><strong>Integration</strong></summary>

```
BUILD SUCCESSFUL in 58s
48 actionable tasks: 23 executed, 25 up-to-date
```

</details>

<details><summary><strong>Acceptance</strong></summary>

```
> Task :airbyte-integrations:connectors:source-dynamodb:sourceAcceptanceTest
============================= test session starts ==============================
platform linux -- Python 3.9.11, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /test_input
plugins: timeout-1.4.2, sugar-0.9.5
collected 21 items / 1 skipped / 20 selected

test_core.py ....................                                        [ 95%]
test_full_refresh.py .                                                   [100%]
```

</details>

### CODE REVIEW

**Comment by alafanechere** [comment]
Hey @Tomperez98 , thank you for this contribution! Could you please share the acceptance test output by running: 
```
./gradlew :airbyte-integrations:connectors:source-dynamodb:integrationTest
```

**Comment by Tomperez98** [comment]
Sure. I'll run that.

Still, the source connector is a work in progress. So no tests have been implemented yet.

**Comment by alafanechere** [comment]
> Still, the source connector is a work in progress. So no tests have been implemented yet.

All right, I'm converting the PR to draft then. Feel free to convert it back to ready for review once you want a review from our side.

**Comment by alafanechere** [comment]
> Still, the source connector is a work in progress. So no tests have been implemented yet.

All right, I'm converting the PR to draft then. Feel free to convert it back to ready for review once you want a review from our side.

**Comment by Tomperez98** [comment]
@alafanechere Sure!


I already able to run `pytest -s integration_tests.acceptance_tests` successfully. 

But when running `./gradlew :airbyte-integrations:connectors:source-dynamodb:integrationTest` I'm getting this error, any idea what might be wrong? I have almost no experience with Java.

```
* Where:
Build file '/Users/tomasperez/Documents/Factored/airbyte-dynamodb-connector/build.gradle' line: 197

* What went wrong:
A problem occurred evaluating root project 'airbyte'.
> java.lang.UnsupportedClassVersionError: com/bmuschko/gradle/docker/DockerRemoteApiPlugin has been compiled by a more recent version of the Java Runtime (class file version 55.0), this version of the Java Runtime only recognizes class file versions up to 52.0

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org
```

**Comment by marcosmarxm** [comment]
@Tomperez98 you need to use Java 17

**Comment by Tomperez98** [comment]
@marcosmarxm @alafanechere 
```
> Task :airbyte-integrations:bases:source-acceptance-test:unitTest
Name                                                 Stmts   Miss  Cover
------------------------------------------------------------------------
source_acceptance_test/utils/__init__.py                 6      0   100%
source_acceptance_test/tests/__init__.py                 4      0   100%
source_acceptance_test/__init__.py                       2      0   100%
source_acceptance_test/tests/test_full_refresh.py       52      2    96%
source_acceptance_test/utils/asserts.py                 37      2    95%
source_acceptance_test/config.py                        77      6    92%
source_acceptance_test/utils/json_schema_helper.py     105     13    88%
source_acceptance_test/tests/test_incremental.py       121     25    79%
source_acceptance_test/utils/common.py                  80     17    79%
source_acceptance_test/tests/test_core.py              294    106    64%
source_acceptance_test/utils/compare.py                 62     23    63%
source_acceptance_test/base.py                          10      4    60%
source_acceptance_test/utils/connector_runner.py       110     48    56%
------------------------------------------------------------------------
TOTAL                                                  960    246    74%
```

```
> Task :airbyte-integrations:connectors:source-dynamodb:sourceAcceptanceTest
============================= test session starts ==============================
platform linux -- Python 3.9.11, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /test_input
plugins: timeout-1.4.2, sugar-0.9.5
collected 20 items / 2 skipped / 18 selected

test_core.py ....................                                        [100%]
```

File: airbyte-integrations/connectors/source-dynamodb/acceptance-test-config.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  #        extra_fields: no
+  #        exact_order: no
+  #        extra_records: yes
+  # incremental: # TODO if your connector does not implement incremental sync, remove this block
+  #   - config_path: "secrets/config.json"
+  #     configured_catalog_path: "integration_tests/configured_catalog.json"
+  #     future_state_path: "integration_tests/abnormal_state.json"
+  # full_refresh:
+  #   - config_path: "secrets/config.json"
+  #     configured_catalog_path: "integration_tests/configured_catalog.json"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-dynamodb/acceptance-test-config.yml:30]: You need to re-enable this and run integration test again.

File: airbyte-integrations/connectors/source-dynamodb/source_dynamodb/constants.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+MAX_ITEMS_PAGINATOR: int = 50
+MAX_PAGES_PAGINATOR: int = 10
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-dynamodb/source_dynamodb/constants.py:6]: You're using this only in one file I think it's better move to the file to improve readability.

File: airbyte-integrations/connectors/source-dynamodb/source_dynamodb/typing.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+from typing import TypedDict
+
+
+class Spec(TypedDict):
+    aws_access_key_id: str
+    aws_secret_access_key: str
+    region_name: str
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-dynamodb/source_dynamodb/typing.py:11]: Remove it and check how other connectors implement this.

File: airbyte-integrations/connectors/source-dynamodb/unit_tests/unit_test.py [hunk 4]
```diff
@@ -0,0 +1,7 @@
+#
+# Copyright (c) 2022 Airbyte, Inc., all rights reserved.
+#
+
+
+def test_example_method():
+    assert True
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-dynamodb/unit_tests/unit_test.py:7]: This is not testing anything please add unit test for the functions

File: airbyte-integrations/connectors/source-dynamodb/run-methods.sh [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env sh
+
+echo Spec
+docker run --rm -v $(pwd)/secrets:/secrets airbyte/source-dynamodb:dev spec
+echo Check
+docker run --rm -v $(pwd)/secrets:/secrets airbyte/source-dynamodb:dev check --config /secrets/config.json
+echo Discover
+docker run --rm -v $(pwd)/secrets:/secrets airbyte/source-dynamodb:dev discover --config /secrets/config.json
+echo Read
+docker run --rm -v $(pwd)/secrets:/secrets -v $(pwd)/integration_tests:/integration_tests airbyte/source-dynamodb:dev read --config /secrets/config.json --catalog /integration_tests/configured_catalog.json
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-dynamodb/run-methods.sh:10]: Remove this file.

File: airbyte-integrations/connectors/source-dynamodb/integration_tests/sample_config.json [hunk 6]
```diff
@@ -0,0 +1,3 @@
+{
+  "fix-me": "TODO"
+}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-dynamodb/integration_tests/sample_config.json:3]: Check other connectors and update this file.

File: airbyte-integrations/connectors/source-dynamodb/integration_tests/invalid_config.json [hunk 7]
```diff
@@ -0,0 +1,5 @@
+{
+  "aws_access_key_id": "not_an_access_key_id",
+  "aws_secret_access_key": "not_a_secret_access_key",
+  "region_name": "us-east-1"
+}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-dynamodb/integration_tests/invalid_config.json:5]: EOF

File: airbyte-integrations/connectors/source-dynamodb/integration_tests/configured_catalog.json [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+          "full_refresh"
+        ],
+        "source_defined_cursor": false,
+        "default_cursor_field": []
+      },
+      "sync_mode": "full_refresh",
+      "destination_sync_mode": "overwrite"
+    }
+  ]
+}
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-dynamodb/integration_tests/configured_catalog.json:26]: EOF

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
Some comments, it's missing the document in docs/integrations/source

File: airbyte-integrations/connectors/source-dynamodb/acceptance-test-config.yml (hunk 1 above)
Comment (Tomperez98) [reply airbyte-integrations/connectors/source-dynamodb/acceptance-test-config.yml:30]: work in progress...

File: airbyte-integrations/connectors/source-dynamodb/run-methods.sh (hunk 5 above)
Comment (Tomperez98) [reply airbyte-integrations/connectors/source-dynamodb/run-methods.sh:10]: Done

**Comment by Tomperez98** [comment]
@marcosmarxm Hi Marcos. I just updated the code solving all your previous comments.

Ready for review. Thanks

@alafanechere

**Comment by Tomperez98** [comment]
All development and tests run locally against a customized DynamoDB Docker Image.

Should I use a DynamoDB table in the cloud? (Thinking about CI/CD possible problems)

**Comment by marcosmarxm** [comment]
> All development and tests run locally against a customized DynamoDB Docker Image.
> 
> Should I use a DynamoDB table in the cloud? (Thinking about CI/CD possible problems)

Would be easier to us test using AWS Dynamodb in cloud

**Comment by Tomperez98** [comment]
>

**Comment by Tomperez98** [comment]
> > All development and tests run locally against a customized DynamoDB Docker Image.
> > Should I use a DynamoDB table in the cloud? (Thinking about CI/CD possible problems)
> 
> Would be easier to us test using AWS Dynamodb in cloud

I agree.

**Comment by marcosmarxm** [comment]
@sajarin can you review this contribution? 
@Tomperez98 I'll be on vacation until 26-jul

**Comment by Tomperez98** [comment]
@sajarin Hi, hope u are doing great!

Please let me know how can I improve the PR

**Comment by marcosmarxm** [comment]
@Tomperez98 The connector it's not working with the parameter `table_prefix_name` can you chceck that?

**Comment by Tomperez98** [comment]
Sure. I'll take a look early next week

**Comment by marcosmarxm** [comment]
@Tomperez98 do you want to continue the work here?

**Comment by Tomperez98** [comment]
@marcosmarxm

**Comment by edreispi** [comment]
@marcosmarxm I downloaded the fork of the repo @ this PR and tried to run the gradle build but it gets stuck at 85%. The repo has been synced with the main one after a few modifications. Has there been a change in the gradle config that requires an additional file or something when testing? Is there a minimum test coverage required for the integration test to complete?

Tried to run `./gradlew :airbyte-integrations:connectors:source-dynamodb:integrationTest --info` in the main airbyte directory.

Last output:

```
> Task :airbyte-integrations:connectors:source-dynamodb:build
Skipping task ':airbyte-integrations:connectors:source-dynamodb:build' as it has no actions.
:airbyte-integrations:connectors:source-dynamodb:build (Thread[Execution worker for ':' Thread 7,5,main]) completed. Took 0.0 secs.
<===========--> 85% EXECUTING [12m 46s]
> IDLE
> IDLE
> IDLE
> IDLE
> :airbyte-integrations:bases:source-acceptance-test:_unitTestCoverage
> IDLE
> IDLE
> IDLE
```

**Comment by edreispi** [comment]
Hello @marcosmarxm 

Picked up this PR after some modifications and tried to run the gradle build locally, this was the output from the tests:
```
============================= test session starts ==============================
platform linux -- Python 3.9.11, pytest-6.2.5, py-1.11.0, pluggy-1.0.0
rootdir: /test_input
plugins: cov-3.0.0, hypothesis-6.54.4, mock-3.6.1, requests-mock-1.9.3, sugar-0.9.5, timeout-1.4.2
collected 26 items / 1 skipped / 25 selected

test_core.py ..........sF.........Fs..                                   [ 96%]
test_full_refresh.py .                                                   [100%]


=================================== FAILURES ===================================
_____________ TestSpec.test_additional_properties_is_true[inputs0] _____________

self = <source_acceptance_test.tests.test_core.TestSpec object at 0x7fe4c078cd90>
actual_connector_spec = ConnectorSpecification(documentationUrl=AnyUrl('https://docsurl.com', scheme='https', host='docsurl.com', tld='com', h...malization=False, supportsDBT=False, supported_destination_sync_modes=None, authSpecification=None, advanced_auth=None)

    def test_additional_properties_is_true(self, actual_connector_spec: ConnectorSpecification):
        """Check that value of the "additionalProperties" field is always true.
        A spec declaring "additionalProperties": false introduces the risk of accidental breaking changes.
        Specifically, when removing a property from the spec, existing connector configs will no longer be valid.
        False value introduces the risk of accidental breaking changes.
        Read https://github.com/airbytehq/airbyte/issues/14196 for more details"""
        additional_properties_values = find_all_values_for_key_in_schema(
            actual_connector_spec.connectionSpecification, "additionalProperties"
        )
        if additional_properties_values:
>           assert all(
                [additional_properties_value is True for additional_properties_value in additional_properties_values]
            ), "When set, additionalProperties field value must be true for backward compatibility."
E           AssertionError: When set, additionalProperties field value must be true for backward compatibility.
E           assert False
E            +  where False = all([False])

/usr/local/lib/python3.9/site-packages/source_acceptance_test/tests/test_core.py:194: AssertionError
__________ TestDiscovery.test_additional_properties_is_true[inputs0] ___________

self = <source_acceptance_test.tests.test_core.TestDiscovery object at 0x7fe4c0636d60>
discovered_catalog = {'ProductCatalog': AirbyteStream(name='ProductCatalog', json_schema={'$schema': 'http://json-schema.org/draft-07/schem...ll_refresh'>], source_defined_cursor=None, default_cursor_field=None, source_defined_primary_key=None, namespace=None)}

    def test_additional_properties_is_true(self, discovered_catalog: Mapping[str, Any]):
        """Check that value of the "additionalProperties" field is always true.
        A stream schema declaring "additionalProperties": false introduces the risk of accidental breaking changes.
        Specifically, when removing a property from the stream schema, existing connector catalog will no longer be valid.
        False value introduces the risk of accidental breaking changes.
        Read https://github.com/airbytehq/airbyte/issues/14196 for more details"""
        for stream in discovered_catalog.values():
            additional_properties_values = list(find_all_values_for_key_in_schema(stream.json_schema, "additionalProperties"))
            if additional_properties_values:
>               assert all(
                    [additional_properties_value is True for additional_properties_value in additional_properties_values]
                ), "When set, additionalProperties field value must be true for backward compatibility."
E               AssertionError: When set, additionalProperties field value must be true for backward compatibility.
E               assert False
E                +  where False = all([False])

/usr/local/lib/python3.9/site-packages/source_acceptance_test/tests/test_core.py:306: AssertionError
=========================== short test summary info ============================
FAILED test_core.py::TestSpec::test_additional_properties_is_true[inputs0] - ...
FAILED test_core.py::TestDiscovery::test_additional_properties_is_true[inputs0]
SKIPPED [1] ../usr/local/lib/python3.9/site-packages/source_acceptance_test/plugin.py:60: Skipping TestIncremental.test_two_sequential_reads because not found in the config
SKIPPED [1] ../usr/local/lib/python3.9/site-packages/source_acceptance_test/tests/test_core.py:51: The previous connector image could not be retrieved.
SKIPPED [1] ../usr/local/lib/python3.9/site-packages/source_acceptance_test/tests/test_core.py:229: The previous connector image could not be retrieved.
=================== 2 failed, 22 passed, 3 skipped in 46.32s ===================
```

Can you help point us in the right direction to fix this backward compatibility issue? Is it part of the connector config?

**Comment by marcosmarxm** [comment]
You must change all the schemas and spec the `additionalProperties` to `true`

**Pull request closed without merging**